src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,424 @@
+        }
+
+        Task<IEnumerable<Guid>> ITransactionConfirmationWatcherHandler<Guid>.CreateContextsAsync(Transaction tx, CancellationToken cancellationToken)
+        {
+            this.timerLock.EnterReadLock();
+
+            try
+            {
+                if (this.timers.TryGetValue(tx.GetHash(), out var txTimers))
+                {
+                    return Task.FromResult<IEnumerable<Guid>>(txTimers.Select(t => t.Key).ToList());
+                }
+            }
+            finally
+            {
+                this.timerLock.ExitReadLock();
+            }
+
+            return Task.FromResult(Enumerable.Empty<Guid>());
+        }
+
+        async Task<bool> IConfirmationWatcherHandler<TransactionWatch<Guid>, Guid>.ConfirmationUpdateAsync(TransactionWatch<Guid> watch, int confirmation, ConfirmationType type, CancellationToken cancellationToken)
+        {
+            switch (type)
+            {
+            case ConfirmationType.Unconfirming:
+                if (confirmation == 1)
+                {
+                    return false;
+                }
+                break;
+
+            case ConfirmationType.Confirmed:
+                if (confirmation == 1)
+                {
+                    return false;
+                }
+
+                this.timerLock.EnterUpgradeableReadLock();
+                try
+                {
+                    var requiredConfirmations = this.timers[watch.TransactionId][watch.Context].Item2;
+
+                    if (confirmation == requiredConfirmations)
+                    {
+                        await ConfirmAsync(watch, cancellationToken);
+                        return true;
+                    }
+                }
+                finally
+                {
+                    this.timerLock.ExitUpgradeableReadLock();
+                }
+
+                break;
+
+            default:
+                throw new NotSupportedException($"{nameof(ConfirmationType)} is not supported");
+            }
+
+            return false;
+        }
+
+        Task<IEnumerable<TransactionWatch<Guid>>> IConfirmationWatcherHandler<TransactionWatch<Guid>, Guid>.GetCurrentWatchesAsync(CancellationToken cancellationToken)
+        {
+            var watches = new Collection<TransactionWatch<Guid>>();
+            foreach (var watch in this.watches)
+            {
+                watches.Add(watch.Value);
+            }
+
+            return Task.FromResult(watches.AsEnumerable());
+        }
+
+        async Task IWatcherHandler<TransactionWatch<Guid>, Guid>.AddWatchesAsync(IEnumerable<TransactionWatch<Guid>> watches, CancellationToken cancellationToken)
+        {
+            if (watches == null)
+            {
+                throw new ArgumentNullException(nameof(watches));
+            }
+
+            foreach (var watch in watches)
+            {
+                await StopTimer(watch.TransactionId, watch.Context);
```

Comment:
We should check if timer fully successfully stopped due to it is possible that it already timeout before reached this line.